### ISSUE TITLE
Spread of set into object not detected as type error

### ISSUE THREAD

**Issue by cmtm** [microsoft/TypeScript#53747]
# Bug Report

A spread of a Set into an object is assignable to a Set. This shouldn't be allowed, because the object isn't a Set. tsc should be able to detect this at compile time.

### 🔎 Search Terms

set spread

### 🕗 Version & Regression Information

This is the behavior in the version I tried (3.3.3, 4.6.2, 5.0.4, 5.1.0-dev.20230411), and I reviewed the FAQ for entries about [Common "Bugs" That Aren't Bugs](https://github.com/microsoft/TypeScript/wiki/FAQ#common-bugs-that-arent-bugs)

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Sets and spreads

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/MYewdgzgLgBAtgTwPoQKZQIwC4YGV0A8YArnAEaoBOAfDALwxioDue6AFAJQDcAsAFAB6QTAAqAC1QwoCAA5SQAM3jI0UAEwwAlhBgBDYFGJ6ANiYQwA8mQBWqQwDoBw6RGAwAJiFQQwAclgPdHtYKHEdABp9MA9pcLAAa10tKD9dPRgAA3woIlIKGkynflBIWEQUdHUcHLzyKloGAG8HVoq1DABfPn5nEQkdGGYtMxhFPRH9WEpiMCgtOCkKYD1iNBVKjW1fAP02KGL2qodxPQh2dR4BIA)

### 💻 Code

```ts
const my_set1: Set<number> = new Set();
// The type of my_set2 is actually Object.
// tsc doesn't detect this, and thinks it's a `Set<number>`.
const my_set2: Set<number> = {...my_set1};

// This will fail at runtime because my_set2 isn't a Set.
my_set2.has(2);
```

### 🙁 Actual behavior

tsc doesn't detect the type error.

### 🙂 Expected behavior

tsc should detect the type error.

**Comment by RyanCavanaugh** [comment]
Yeah, `{ ...x }` is always assumed to be the same type of `x`. Non-enumerability isn't tracked anyway, and expanding it out into `{ a: x.a, b: x.b, c: x.c }` etc breaks a bunch of scenarios that people rightly expect to work.

**Comment by MartinJohns** [comment]
> Non-enumerability isn't tracked anyway

Related: #9726
